### PR TITLE
multi_step_cov_estimator: added 'default_initial_pose' parameter

### PR DESCRIPTION
**Pull request by dgossow**
This will replace the previously used value [0 0 0 0 0 0] as initial pose guess by an (optional) parameter placed in the system.yaml. On dewey, the calibration woudn't converge otherwise.

Not sure about this. Maybe this should go into the 'checkerboards' section.

Example system.yaml:

```
base_link: base_link

sensors:

  chains:
    torso_chain:
      root: base_link
      tip: head_tilt_link
      cov:
       joint_angles: [0.01, 0.01, 0.01, 0.01, 0.01]
      gearing: [1.0, 1.0, 1.0, 1.0, 1.0]

  rectified_cams:
    head_camera:
      chain_id: torso_chain
      frame_id: head_camera_rgb_optical_frame
      baseline_shift: 0.0
      baseline_rgbd: 0.075
      f_shift: 0.0
      cx_shift: 0.0
      cy_shift: 0.0
      cov: {u: 0.25, v: 0.25, x: 0.25}
    base_camera:
      chain_id: NONE
      frame_id: base_camera_rgb_optical_frame
      baseline_shift: 0.0
      baseline_rgbd: 0.075
      f_shift: 0.0
      cx_shift: 0.0
      cy_shift: 0.0
      cov: {u: 0.25, v: 0.25, x: 0.25}

  tilting_lasers: {}

checkerboards:
  cb_7x6:
    corners_x: 7
    corners_y: 6
    spacing_x: 0.108
    spacing_y: 0.108

transforms:
  dummy_transform:   [ 0, 0, 0, 0, 0, 0 ]

default_initial_pose: [ 3, 0, 0, 0, 0, 0 ]
```

### CODE REVIEW

File: calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py [hunk 1]
```diff
@@ -49,6 +49,7 @@
 import os
 
 from numpy import matrix
+from numpy import multiply
```
Comment (ahendrix) [inline calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py]: This looks like a stray import; does it belong in a different patch?

**Comment by ahendrix** [comment]
From a configuration perspective, I think this belongs in the sample configuration, since each sample potentially expects the checkerboard in a different position.

I suspect doing that properly will be a real pain to implement, since you'll probably have to pass the pose guesses through the bag file to the estimator.

File: calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py [hunk 2]
```diff
@@ -254,6 +255,9 @@ def update_urdf(urdf, calibrated_params):
         previous_pose_guesses = numpy.array(yaml.load(config['initial_poses']))
     else:
         previous_pose_guesses = numpy.zeros([msg_count,6])
+        if 'default_initial_pose' in config.keys():
+            for p in range(msg_count):
+                previous_pose_guesses[p,] = config['default_initial_pose']
```
Comment (ahendrix) [inline calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py]: What happens if the user specifies a default pose array with an insufficient number of elements?

**Comment by mikeferguson** [comment]
So this patch is really a huge hack. The more proper way (and what Austin alludes to above with checkerboards in different positions) is really to set a "best sensor", and have the system set the starting guess pose of each checkerboard sample to the value obtained via reprojection of the checkerboard based on a particular sample data. (for instance, on the PR2, you might suggest the tilt laser is the "best sensor" and it will use that to reproject where all poses are.

This of course does lead to a problem: if your best sensor is the arm, well, floating checkerboards are screwed. In this case, perhaps we can rename this parameter to "default_floating_initial_pose" or similar, and then we have implemented the simple version, and can later come back and make the more complex "best sensor" approach happen.

Which actually brings me to another thing: "best sensor" should probably be an array, so that you could say for the PR2 that left_arm then right_arm, and then when left_arm fails for a checkerboard held in the right_arm, you can fall back to right_arm. Yeah, that sounds like quite a bit of code. For now, let's just get "floating" into the name of this parameter and merge it (perhaps also put a comment with reference to this thread of future "TODO: add alternate methods of defining default poses guesses")

**Comment by dgossow** [comment]
Made the requested changes.

File: calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py [hunk 3]
```diff
@@ -254,6 +254,16 @@ def update_urdf(urdf, calibrated_params):
         previous_pose_guesses = numpy.array(yaml.load(config['initial_poses']))
     else:
         previous_pose_guesses = numpy.zeros([msg_count,6])
+        
+        # TODO: add alternate methods of defining default poses guesses
+        # See https://github.com/ros-perception/calibration/pull/9
+        if 'default_floating_initial_pose' in config.keys():
+            default_pose = config['default_floating_initial_pose']
+            if len(default_pose) != 6:
+                print "The 'default_initial_pose' parameter has", len(default_pose), "elements, but it should have 6!"
```
Comment (ahendrix) [inline calibration_estimation/src/calibration_estimation/multi_step_cov_estimator.py]: default_floating_initial_pose?

**Comment by dgossow** [comment]
fixed the default_floating_initial_pose thing
